Experimental Implementation of Consul Storage for Namerd

Ashald opened the pull request:
Although there are no integration tests I verified that it works with local Consul installation.

Ashald commented:
I'm a super-newby in Scala and I believe this code make use of a lot of refactoring and optimizations. I'd highly appreciate code review and change requests so I can bring this code to the level of quality you will agree to merge into the upstream.

Thanks!

samuel commented:
This is awesome.

The one thing it seems though is that this PR is using the KV store rather than Consul's service registration. Using the services rather than kv would probably be better as it's the standard way to handle this with Consul as it then allows using the health checks and DNS resolution.

https://www.consul.io/docs/agent/http/catalog.html

Ashald commented:
@samuel The Consul-enabled `io.l5d.consul` namer in already implemented in Linkerd https://linkerd.io/doc/0.7.0/linkerd/namer/ and it uses Catalog API.

This PR targets the Dtab storage interface rather than namer. And it cannot be implemented with the Catalog API.

samuel commented:
Ahh, cool, my bad. Thanks for the link.

Ashald commented:
@samuel you're welcome ;)

adleong commented:
From giving this a read-through, it looks really good.  I have a bunch of style comments and suggestions but overall it looks great.

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -44,7 +49,7 @@ package object v1 {
   }
 
-  object Api {
+  object CatalogApi {
```
Could we refactor this into an `Api` base trait and have `CatalogApi` and `KvApi` inherit from it?  Could also split these into 3 files (Api, CatalogApi, KvApi).

Ashald replied:
You mean classes, right? Or objects as well? I may miss something but with regards to objects I see only 1 'generalization option' - add a type parameter to the base one and implement `apply` there...

adleong replied:
I don't think we would factor up `apply`, but definitely `infiniteRetryFilter` and `retryClient` could be factored up.

```
trait Api {
  ...
}

object CatalogApi extends Api {
  ...
}

object KvApi extends Api {
  ...
}
```

Ashald commented:
@adleong once you have some time, can you please share your comments and suggestions so I can improve the code?

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -121,6 +126,128 @@ package object v1 {
+
+    val kvPrefix = s"$uriPrefix/kv"
+
+    private[this] val infiniteRetryFilter = new RetryFilter[http.Request, http.Response](
```
infiniteRetryFilter and retryClient could be factored into the `Api` common trait

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -121,6 +126,128 @@ package object v1 {
+        "dc" -> datacenter
+      )
+      Trace.letClear(retryClient(retry)(req)).flatMap {
+        case rsp if rsp.status == http.Status.NotFound => Future.exception(NotFound(rsp))
```
Can all the status code checking be moved into `Indexed.mk`?

Ideally this could just be 
``` 
Trace.letClear(...).flatMap { rsp =>
  Future.const(Indexed.mk[Seq[String]](rsp)
}
```

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -121,6 +126,128 @@ package object v1 {
+
+    // https://www.consul.io/docs/agent/http/kv.html#single
+    def list(
+      path: String,
```
might be worth noting in a comment/scaladoc that path must begin with a slash (I assume?)

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -121,6 +126,128 @@ package object v1 {
+        "dc" -> datacenter
+      )
+
+      Trace.letClear(retryClient(retry)(req)) flatMap {
```
style nit (here and elsewhere): avoid infix notation for methods that take a block.

e.g.
do
`foo.flatMap {`

not
`foo flatMap {`

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -121,6 +126,128 @@ package object v1 {
+
+      Trace.letClear(retryClient(retry)(req)) flatMap {
+        case rsp if rsp.status == http.Status.Ok =>
+          Future.value(Indexed[String](rsp.getContentString(), rsp.headerMap.get(Headers.Index)))
```
why not use `Indexed.mk`?

Ashald replied:
`Indexed.mk` tries to parse data with JSON - this endpoint with given parameters returns value as is. If it's an arbitrary string JSON will fail since not everything can be a top-level object in JSON. But I like the idea...

adleong replied:
what about creating a `Indexed.mkString` that takes a `http.Response` and returns a `Try[Indexed[String]]`.  It would share the same status code and index handling with `Indexed.mk` but would use the raw string instead of json parsing it.  what do you think?

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -121,6 +126,128 @@ package object v1 {
+      req.setContentString(value)
+      Trace.letClear(retryClient(retry)(req)).flatMap {
+        case rsp if rsp.status == http.Status.Ok =>
+          Future.const(readJson[Boolean](rsp.content))
```
Could use `Indexed.mk` here too.  Of course, the return type of this method would change to `Future[Indexed[Boolean]]` and the index would always be `None`, but I think that's okay.

adleong commented on consul/src/main/scala/io/buoyant/consul/v1.scala:
```diff
@@ -132,6 +259,7 @@ package object v1 {
         readJson[T](rsp.content).map { t =>
           Indexed[T](t, rsp.headerMap.get(Headers.Index))
         }
+      case http.Status.Forbidden => throw Forbidden(rsp)
```
could move all status code handling to here

adleong commented on consul/src/test/scala/io/buoyant/consul/KvApiTest.scala:
```diff
@@ -0,0 +1,225 @@
+  test("list returns an indexed seq of key names") {
+    val service = stubService(listBuf)
+
+    val result = await(KvApi(service).list("foo/"))
```
it doesn't really impact this test, but I think path needs to start with a slash or else you'll get some weird uris.

adleong commented on consul/src/test/scala/io/buoyant/consul/KvApiTest.scala:
```diff
@@ -0,0 +1,225 @@
+      lastUri = req.uri
+      Future.value(rsp)
+    }
+    try {
```
this could be

`assertThrows[NotFound](await(...))`

(you have to import `io.buoyant.test.Exceptions`)

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,102 @@
+import com.twitter.util._
+import io.buoyant.consul.NotFound
+import io.buoyant.consul.v1._
+
```
don't need this blank line

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStoreInitializer.scala:
```diff
@@ -0,0 +1,48 @@
+import io.buoyant.config.types.Port
+import io.buoyant.consul.SetHostFilter
+import io.buoyant.consul.v1.KvApi
+
```
don't need this blank line

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,102 @@
+              val namespaces = result.value.map(namespace).toSet
+              updates() = Activity.Ok(namespaces)
+              cycle(result.index)
+            case Throw(e) =>
```
I think we need to handle NotFounds here and turn them into an empty namespace list

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,102 @@
+
+  def update(ns: Ns, dtab: Dtab, version: Version): Future[Unit] = {
+    val Buf.Utf8(vstr) = version
+    api.put(s"${root.show}/$ns", dtab.show, cas = Some(vstr)) flatMap { result =>
```
consul throws a 500 if the cas value isn't a number.  we should check here that vstr can be parsed as a number, and immediately return a version mismatch exception if not

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,102 @@
+              cycle(result.index)
+            case Throw(e: NotFound) =>
+              updates() = Activity.Ok(None)
+              cycle(None)
```
in the case of a NotFound, we still want to extract the `X-Consul-Index` and set the index, otherwise we'll just send keep sending requests here.

Ashald commented:
@adleong I hope I didn't miss anything and addressed all your comments. Can you please take another look? Although now there are so much changes I believe it makes more sense to look just at the code under `consul` and `namerd/storage/consul` rather than at the diff.

adleong commented on consul/src/main/scala/io/buoyant/consul/v1/BaseApi.scala:
```diff
@@ -0,0 +1,76 @@
+  mapper.registerModule(DefaultScalaModule)
+  mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+
+  def parseJson[T: Manifest](buf: Buf): T = {
```
I think this should still be wrapped in a `Try` like before, to help protect against malformed responses

adleong commented on consul/src/main/scala/io/buoyant/consul/v1/BaseApi.scala:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  private[v1] def execute[T: Manifest](req: http.Request, retry: Boolean): Future[Indexed[T]] = {
+    Trace.letClear(getClient(retry)(req)).map { rsp =>
```
I think the idiomatic way to mesh this with `parseJson` returning `Try` would be
```
for {
  rsp <- Trace.letClear(getClient(retry)(req))
  value <- Future.const(parseJson[T](rsp.content))
} yield Indexed[T](value, rsp.headerMap.get(Headers.Index))
```

adleong commented on consul/src/main/scala/io/buoyant/consul/v1/KvApi.scala:
```diff
@@ -0,0 +1,114 @@
+      "index" -> blockingIndex,
+      "dc" -> datacenter
+    )
+    execute[Seq[ConsulKey]](req, retry).map { idx =>
```
by using the `executeRaw` method I suggest above, we could go back to using raw mode here, which seems nicer

Ashald replied:
You mean an alternative version of `execute` that doesn't require de-serializng? Ok.

adleong commented on consul/src/main/scala/io/buoyant/consul/v1/package.scala:
```diff
@@ -0,0 +1,31 @@
+  case class NotFound(rsp: http.Response) extends ConsulApiError
+  case class Forbidden(rsp: http.Response) extends ConsulApiError
+
+  private[v1] val apiErrorFilter = new SimpleFilter[http.Request, http.Response] {
```
good call making this a filter, I like it

Ashald replied:
Thanks! :)

adleong commented on consul/src/main/scala/io/buoyant/consul/v1/package.scala:
```diff
@@ -0,0 +1,31 @@
+      service(request).map { response: http.Response =>
+        response.status match {
+          case http.Status.Ok => response
+          case http.Status.Forbidden => throw Forbidden(response)
```
prefer to use `Future.exception` instead of `throw`:

```
service(request).flatMap { ...
  ...
    case http.Status.Ok => Future.value(response)
    case http.Status.Forbidden => Future.exception(Forbidden(response))
    ...
```

adleong commented on consul/src/test/scala/io/buoyant/consul/v1/CatalogApiTest.scala:
```diff
@@ -99,11 +95,8 @@ class ApiTest extends FunSuite with Awaits {
-    } catch {
-      case e: UnexpectedResponse => assert(true)
-    }
+    assertThrows[UnexpectedResponse](
```
nice, thanks

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+import io.buoyant.namerd.DtabStore.{DtabNamespaceAlreadyExistsException, DtabNamespaceDoesNotExistException}
+
+class ConsulDtabStore(api: KvApi, root: Path) extends DtabStore {
+  import DtabStore.{DtabVersionMismatchException, Version}
```
use absolute imports and put them at the top of the file

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+
+      def cycle(index: Option[String]): Future[Unit] =
+        if (running)
+          api.list(s"${root.show}/", blockingIndex = index) transform {
```
`).transform {`

Ashald replied:
Ooops... :)

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+          Future.Unit
+      val pending = cycle(None)
+
+      Closable make { _ =>
```
`Closable.make`

Ashald replied:
Another ooops.. :)

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  def create(ns: Ns, dtab: Dtab): Future[Unit] =
+    api.put(s"${root.show}/$ns", dtab.show, cas = Some("0")) flatMap { result =>
```
`).flatMap {`

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+
+  def delete(ns: Ns): Future[Unit] = {
+    val key = s"${root.show}/$ns"
+    api.get(key) transform {
```
`).transform {`

Ashald replied:
Argh, I'm angry at myself

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+    val key = s"${root.show}/$ns"
+    api.get(key) transform {
+      case Return(_) => api.delete(key).unit
+      case Throw(e: NotFound) => Future.exception(new DtabNamespaceDoesNotExistException(ns))
```
as far as I can tell, consul seems to always return 200 and "true", even if the key doesn't exist.  *shrug*

Ashald replied:
Yeah, that's for `delete`. that's why I added a `get` first so we can raise a `DtabNamespaceDoesNotExistException ` as needed.

adleong replied:
oh, yeah, that makes sense.  good call

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+
+  def update(ns: Ns, dtab: Dtab, version: Version): Future[Unit] = {
+    val Buf.Utf8(vstr) = version
+    if (!vstr.forall(_.isDigit)) Future.exception(new DtabVersionMismatchException)
```
what about `if (Try(vstr.toLong)).isThrow)`?

Ashald replied:
I like your version better - I'm just not sure about consul index boundaries. But it should fit into Long, right? :)

adleong replied:
I think you're missing either a `return Future.exception(...)` or an else block.  Currently the `Future.exception` isn't returned, it's just dropped.

Ashald replied:
Thanks!

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+    val Buf.Utf8(vstr) = version
+    if (!vstr.forall(_.isDigit)) Future.exception(new DtabVersionMismatchException)
+
+    api.put(s"${root.show}/$ns", dtab.show, cas = Some(vstr)) flatMap { result =>
```
`).flatMap {`

Ashald replied:
I'm sorry. :(

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+
+      def cycle(index: Option[String]): Future[Unit] =
+        if (running)
+          api.get(key, blockingIndex = index) transform {
```
`).transform {`

Ashald replied:
I'm so sorry. :(

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala:
```diff
@@ -0,0 +1,106 @@
+          Future.Unit
+      val pending = cycle(None)
+
+      Closable make { _ =>
```
`Closable.make {`

Ashald replied:
I'm super sorry. :(

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStoreInitializer.scala:
```diff
@@ -0,0 +1,47 @@
+
+  @JsonIgnore
+  override def mkDtabStore: DtabStore = {
+    val serviceHost = host getOrElse DefaultHost
```
`host.getOrElse(DefaultHost)`

Ashald replied:
I just copied it from `etcd`. :D of course will fix. :)

adleong commented on namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStoreInitializer.scala:
```diff
@@ -0,0 +1,47 @@
+  @JsonIgnore
+  override def mkDtabStore: DtabStore = {
+    val serviceHost = host getOrElse DefaultHost
+    val servicePort = (port getOrElse DefaultPort).port
```
`port.getOrElse(DefaultPort)`

adleong commented on namerd/storage/consul/src/test/scala/io/buoyant/namerd/storage/consul/EtcdConfigTest.scala:
```diff
@@ -0,0 +1,29 @@
+package io.buoyant.namerd.storage.consul
```
I think this file has the wrong name

Ashald replied:
Ooops, thanks

Ashald commented:
Another `Try`. :D

adleong commented:
:star: This is great!  Thanks for all your hard work on this, @Ashald!

[![Approved with PullApprove](https://img.shields.io/badge/readability-approved-brightgreen.svg)](https://pullapprove.com/BuoyantIO/linkerd/pull-request/541/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=BuoyantIO/linkerd)

Ashald commented:
Thanks for the help!
